### PR TITLE
Add tests for place holder search

### LINKED ISSUE #15: Test placeholder search
Placeholder search lets the user send a null `query` parameter in order to apply all the ranking rules to the documents in the index without matching them with a specific query, and return the results to the user. It is different to sending an empty string as a query, in which case you won't receive any results. This behavior should be explicitly tested in the SDK.

More info about Placeholder search [here](https://docs.meilisearch.com/guides/main_concepts/search.html#placeholder-search) 

________________________________
UPDATE: 

Placeholder works 👍

Placeholder Search can be used by:
- Using a null query: `q=null`
- Using an empty String: `q=""`

Tests are still missing (both options, null and empty String, should be tested).

### PR DESCRIPTION
**Pull request by luis-valdez**
Closes #15.
In this PR I include to different tests that validate the functionality of place holder search, which consists of making a search a query with an empty string or a null parameter, i.e:
```java
// Both of the following statements return all the documents in the index:
index.search(new SearchRequest(null)); // option 1
index.search(""); // option 2
```

The tests implemented are `testPlaceHolder` and `testPlaceHolderWithLimit`.

I'd appreciate some feedback on the tests because I wasn't sure how atomic the tests should be. If they can be improved in any way, please let me know 😄 !

### CODE REVIEW

File: src/test/java/com/meilisearch/integration/SearchTest.java [hunk 1]
```diff
@@ -315,5 +315,58 @@ public void testSearchMatches() throws Exception {
...
+			UpdateStatus.class
+		);
+
+		index.waitForPendingUpdate(updateInfo.getUpdateId());
+
+		Results res_gson = jsonGson.decode(
+			index.search(""),
+			Results.class
+		);
+		assertEquals(index.search(""), index.search(new SearchRequest(null)));
```
Comment (eskombro) [inline src/test/java/com/meilisearch/integration/SearchTest.java:339]: I would suggest storing the results of `index.search("")` in a Sting variable and then use it in both places (L336 and L339). This won't probably add a noticeable difference now, but seems to me to be a more consistent way of knowing the values are the same always, and also avoid 1 call to the server :)

File: src/test/java/com/meilisearch/integration/SearchTest.java [hunk 2]
```diff
@@ -315,5 +315,58 @@ public void testSearchMatches() throws Exception {
...
+		assertEquals(20, res_gson.hits.length);
+		assertEquals("419704", res_gson.hits[0].getId());
+		assertEquals("Ad Astra", res_gson.hits[0].getTitle());
+		assertEquals("920", res_gson.hits[10].getId());
+		assertEquals("Cars", res_gson.hits[10].getTitle());
+		assertEquals("420817", res_gson.hits[19].getId());
+		assertEquals("Aladdin", res_gson.hits[19].getTitle());
```
Comment (eskombro) [inline src/test/java/com/meilisearch/integration/SearchTest.java:346]: Checking these exact hardcoded values may seem to make sense at first sight, but it can introduce a few problems that may pop up in the future. As MeiliSearch Core continues to evolve, changes in the way the MeiliSearch algorithm works may end up breaking these tests, even if the SDK functionality is still OK (imagine for example that de default order of the ranking rules changes, and so, the order of the results is not the same. But the Placeholder search is still working, right? well according to this test, no. This is the complexity that we are introducing here). So from my POV, we could make this more "generic". We need to know that call works, that it returns results, that we have the default amount of them (20)... but checking the specific ID is always risky. I know we do it sometimes in some other tests, but in this case doesn't seem to be really necessary, WDYT?

**Review by eskombro — COMMENTED** [review]
Good work, thanks! I added two comments, let me know what you think about this :)

**Comment by eskombro** [comment]
Also, I was thinking that it would be interesting to add a test checking that both ways of using placeholder search produce the same result (null and empty string). Am I going too far and this is more of a core test than a SDK test @curquiza ?

**Comment by curquiza** [comment]
> Also, I was thinking that it would be interesting to add a test checking that both ways of using placeholder search produce the same result (null and empty string). Am I going too far and this is more of a core test than a SDK test @curquiza ?

The first test already does it if I'm not wrong:

```java
assertEquals(index.search(""), index.search(new SearchRequest(null)));
```

For me this is enough 🙂

File: src/test/java/com/meilisearch/integration/SearchTest.java (hunk 1 above)
Comment (luis-valdez) [reply src/test/java/com/meilisearch/integration/SearchTest.java:339]: Good catch @eskombro! I didn't notice I was actually making one unnecessary call to the server. I'll implement the changes.

File: src/test/java/com/meilisearch/integration/SearchTest.java (hunk 2 above)
Comment (luis-valdez) [reply src/test/java/com/meilisearch/integration/SearchTest.java:346]: Tbh I was a little confused on how I had to prove that placeholder was working, but now that you mention it, it makes a lot of sense testing it without the hardcoded values.
Checking only the amount of results seems like a good approach. Thank you for the feedback!

**Comment by luis-valdez** [comment]
> > Also, I was thinking that it would be interesting to add a test checking that both ways of using placeholder search produce the same result (null and empty string). Am I going too far and this is more of a core test than a SDK test @curquiza ?
> 
> The first test already does it if I'm not wrong:
> 
> ```java
> assertEquals(index.search(""), index.search(new SearchRequest(null)));
> ```
> 
> For me this is enough 🙂

That's correct. I thought about doing two separate tests, one for the empty string and another one for the null value.
In one hand having two separate tests would make them more atomic, but in the other hand a lot of code would be duplicated.
What do you guys think?

**Comment by curquiza** [comment]
> That's correct. I thought about doing two separate tests, one for the empty string and another one for the null value.
In one hand having two separate tests would make them more atomic, but in the other hand a lot of code would be duplicated.
What do you guys think?

I would rather one test, so let it as it is if @eskombro is ok. If we find out any trouble with this choice in the future, we will change! 🙂

**Review by eskombro — APPROVED** [review]
Good, I think this is enough for testing Placeholder functionality and raise an alert if something is broken. Thanks @luis-valdez !

**Comment by eskombro** [comment]
bors try

**Comment by eskombro** [comment]
bors merge
